File: src/domain/ledger.ts
import { addMoney, subtractMoney, type Halalas } from './money'
import { countsAsIncome, countsAsPersonalExpense } from './entities/economicKind'
import type {
  Obligation,
  ObligationKind,
  PersonAllocation,
  Settlement,
  Transaction,
} from './entities/types'

/**
 * الدفتر — قواعد spec/02-accounting.md كدوال نقية.
 *
 * القاعدة المركزية: **عملية واحدة، مبلغ واحد، طرق عرض متعددة.**
 * الوسوم لا تضاعف المجموع، والفرع يدخل في الأصل مرة واحدة.
 */

export class LedgerError extends Error {}

/* ───────────────────────── المصروف الشخصي ───────────────────────── */

/**
 * نصيب المستخدم من عملية = المبلغ − ما خُصِّص للآخرين كـ **دين** (receivable).
 *
 * التخصيص كـ «هدية» (gift) لا يخصم من المصروف الشخصي:
 * spec/06 «اختيار هدية لشخص ⇒ المصروف كامل؛ لا التزام عليه».
 *
 * مجموع التخصيصات لا يتجاوز قيمة الشراء (spec/03).
 */
export function personalShareOf(
  transaction: Pick<Transaction, 'id' | 'amountMinor' | 'economicKind'>,
  allocations: readonly PersonAllocation[],
): Halalas {
  if (!countsAsPersonalExpense(transaction.economicKind)) return 0

  const mine = allocations.filter((a) => a.transactionId === transaction.id)
  const allocatedTotal = addMoney(...mine.map((a) => a.amountMinor))

  if (allocatedTotal > transaction.amountMinor) {
    throw new LedgerError(
      `مجموع التخصيصات (${allocatedTotal}) أكبر من قيمة العملية (${transaction.amountMinor})`,
    )
  }

  // الدين فقط يخصم؛ الهدية تبقى مصروفك
  const receivable = addMoney(
    ...mine.filter((a) => a.allocationKind === 'receivable').map((a) => a.amountMinor),
  )
  return subtractMoney(transaction.amountMinor, receivable)
}

export interface PeriodTotals {
  incomeMinor: Halalas
  /** المصروف الشخصي: الاستهلاك غير المستبعد، بعد فصل المستحق على الآخرين. */
  personalExpenseMinor: Halalas
  /** المتبقي = الدخل − المصروف. ليس السيولة ولا رصيد البنك (spec/02). */
  remainingMinor: Halalas
  /** المصروف المستبعد من الميزانية — يُعرض منفصلًا ولا يُخفى. */
  excludedExpenseMinor: Halalas
}

/**
 * مجاميع الفترة. المدخل عمليات الفترة فقط — الدالة لا تفلتر بالتاريخ.
 *
 * الاستبعاد من الميزانية لا يلغي خصم المحفظة (spec/02)،
 * لذلك يُطرح من المصروف ويُعرض في خانته المستقلة، ولا يختفي.
 */
export function computePeriodTotals(
  transactions: readonly Transaction[],
  allocations: readonly PersonAllocation[],
): PeriodTotals {
  let income = 0
  let expense = 0
  let excluded = 0

  for (const t of transactions) {
    if (countsAsIncome(t.economicKind)) {
      income = addMoney(income, t.amountMinor)
      continue
    }
    if (!countsAsPersonalExpense(t.economicKind)) continue

    const share = personalShareOf(t, allocations)
    if (t.excludedFromBudget) excluded = addMoney(excluded, share)
    else expense = addMoney(expense, share)
  }

  return {
    incomeMinor: income,
    personalExpenseMinor: expense,
    remainingMinor: subtractMoney(income, expense),
    excludedExpenseMinor: excluded,
  }
}

/* ───────────────────────── الوسوم بلا مضاعفة ───────────────────────── */

/**
 * مجموع عمليات موسومة بوسم معيّن.
 *
 * spec/02: «انضمام جدول الوسوم لا يجوز أن يضاعف SUM».
 * التنفيذ هنا يجمع **العمليات المميزة** بمعرّفها، فوسمان على نفس العملية
 * ينتجان 720 لا 1440، وثلاثة لا ينتجون 2160 (حالة spec/06).
 */
export function sumByTag(
  transactions: readonly Transaction[],
  taggedTransactionIds: readonly string[],
  allocations: readonly PersonAllocation[] = [],
): Halalas {
  const wanted = new Set(taggedTransactionIds)
  const seen = new Set<string>()
  let total = 0
  for (const t of transactions) {
    if (!wanted.has(t.id) || seen.has(t.id)) continue
    seen.add(t.id)
    total = addMoney(total, personalShareOf(t, allocations))
  }
  return total
}

/* ───────────────────────── حسابات الأشخاص ───────────────────────── */

export interface PersonBalance {
  personId: string
  /** «لك عنده» — المتبقي من القروض والمشتريات التي دفعها المستخدم عنه. */
  receivableMinor: Halalas
  /** «له عندك» — قروض على المستخدم. منفصل عن الأمانة داخليًا (spec/02). */
  payableLoanMinor: Halalas
  /** «له عندك» — أمانات. لا تُقاصّ بالقرض ولا تُسدَّد منه ضمنيًا. */
  payableCustodyMinor: Halalas
}

/** المتبقي من التزام = الأصل − مجموع تسوياته. */
export function remainingOfObligation(
  obligation: Obligation,
  settlements: readonly Settlement[],
): Halalas {
  const paid = addMoney(
    ...settlements.filter((s) => s.obligationId === obligation.id).map((s) => s.amountMinor),
  )
  const remaining = subtractMoney(obligation.originalMinor, paid)
  if (remaining < 0) {
    throw new LedgerError(
      `تسويات الالتزام ${obligation.id} تجاوزت أصله — رصيد سالب صامت ممنوع`,
    )
  }
  return remaining
}

/**
 * أرصدة شخص. **لا تقاص تلقائي** بين «لك» و«عليك» (spec/02)،
 * والقرض والأمانة يبقيان منفصلين داخليًا حتى لو عُرض إجماليهما في خانة واحدة.
 */
export function computePersonBalance(
  personId: string,
  obligations: readonly Obligation[],
  settlements: readonly Settlement[],
): PersonBalance {
  const mine = obligations.filter((o) => o.personId === personId)
  const totalOf = (kind: ObligationKind): Halalas =>
    addMoney(...mine.filter((o) => o.kind === kind).map((o) => remainingOfObligation(o, settlements)))

  return {
    personId,
    receivableMinor: totalOf('receivable'),
    payableLoanMinor: totalOf('loan_payable'),
    payableCustodyMinor: totalOf('custody_payable'),
  }
}

export interface SettlementCheck {
  allowed: boolean
  /** سبب الرفض بلغة المستخدم — «رفض مع تفسير» (spec/06). */
  reason?: string
  /** الجزء الذي يُسوّى فعلًا لو قُبل جزئيًا. */
  settledMinor: Halalas
  /** الفائض الذي يحتاج إجراءً صريحًا كأمانة مستقلة. */
  surplusMinor: Halalas
}

/**
 * يتحقق من تسوية قبل تطبيقها.
 *
 * spec/06:
 * - «محاولة رد أمانة 300 والمتاح 200 ⇒ رفض مع تفسير؛
 *    لا السحب من رصيد قرض ضمنيًا»
 * - «سداد أكبر من الدين ⇒ رفض أو تقسيم صريح إلى تسوية وأمانة؛
 *    لا رصيد سالب صامت»
 */
export function checkSettlement(
  obligation: Obligation,
  settlements: readonly Settlement[],
  amountMinor: Halalas,
): SettlementCheck {
  if (amountMinor <= 0) {
    return { allowed: false, reason: 'مبلغ التسوية لازم يكون أكبر من صفر', settledMinor: 0, surplusMinor: 0 }
  }

  const remaining = remainingOfObligation(obligation, settlements)

  if (remaining === 0) {
    return {
      allowed: false,
      reason: 'الالتزام ده متسدد بالكامل. مفيش متبقي يتسوّى',
      settledMinor: 0,
      surplusMinor: amountMinor,
    }
  }

  if (amountMinor <= remaining) {
    return { allowed: true, settledMinor: amountMinor, surplusMinor: 0 }
  }

  // الزيادة لا تُبتلع ولا تُسحب من التزام آخر — تحتاج قرارًا صريحًا
  const kindLabel =
    obligation.kind === 'custody_payable'
      ? 'الأمانة'
      : obligation.kind === 'loan_payable'
        ? 'الدين'
        : 'المستحق'
  return {
    allowed: false,
    reason:
      `المبلغ أكبر من المتبقي. ${kindLabel} المتبقي ${remaining} هللة والمبلغ ${amountMinor}. ` +
      `تقدر تسوّي ${remaining} وتسجّل الباقي ${amountMinor - remaining} كأمانة مستقلة بإجراء صريح.`,
    settledMinor: remaining,
    surplusMinor: subtractMoney(amountMinor, remaining),
  }
}

/* ───────────────────────── التحويل الداخلي والرسوم ───────────────────────── */

export interface GrossSplit {
  principalMinor: Halalas
  feeMinor: Halalas
}

/**
 * يفصل الإجمالي المخصوم إلى أصل ورسوم.
 *
 * spec/06: «رسوم 20 ضمن خصم 1020 ⇒ أصل 1000 + رسوم 20؛ لا إجمالي 1040».
 * أي: الرسوم **لا تُضاف مرة ثانية** إذا كان المخصوم شاملًا لها.
 */
export function splitGrossIntoPrincipalAndFee(
  grossMinor: Halalas,
  feeMinor: Halalas,
): GrossSplit {
  if (feeMinor < 0) throw new LedgerError('الرسوم لا تكون سالبة')
  if (feeMinor > grossMinor) {
    throw new LedgerError(`الرسوم (${feeMinor}) أكبر من الإجمالي المخصوم (${grossMinor})`)
  }
  return { principalMinor: subtractMoney(grossMinor, feeMinor), feeMinor }
}

File: native-app/core/src/commonMain/kotlin/app/masroufy/core/Ledger.kt
package app.masroufy.core

/**
 * الدفتر — نقل `src/domain/ledger.ts` (قواعد spec/02 كدوال نقية).
 * **عملية واحدة، مبلغ واحد، طرق عرض كتير**: الوسوم ما بتضاعفش المجموع.
 */
class LedgerError(message: String) : IllegalStateException(message)

/** نصيب المستخدم = المبلغ − اللي اتخصص للآخرين كـ**دين**. الهدية ما بتخصمش (spec/06). */
fun personalShareOf(transaction: Transaction, allocations: List<PersonAllocation>): Halalas {
    if (!countsAsPersonalExpense(transaction.economicKind)) return 0
    val mine = allocations.filter { it.transactionId == transaction.id }
    val allocatedTotal = sumMoney(mine.map { it.amountMinor })
    if (allocatedTotal > transaction.amountMinor) {
        throw LedgerError(
            uiText(TextKey.LEDGER_ALLOCATIONS_EXCEED, formatMoney(allocatedTotal), formatMoney(transaction.amountMinor)),
        )
    }
    val receivable = sumMoney(mine.filter { it.allocationKind == AllocationKind.RECEIVABLE }.map { it.amountMinor })
    return subtractMoney(transaction.amountMinor, receivable)
}

data class PeriodTotals(
    val incomeMinor: Halalas,
    /** المصروف الشخصي غير المستبعد، بعد فصل اللي على الآخرين. */
    val personalExpenseMinor: Halalas,
    /** الدخل − المصروف. مش السيولة ولا رصيد البنك (spec/02). */
    val remainingMinor: Halalas,
    /** المستبعد من الميزانية — بيتعرض لوحده وما بيختفيش. */
    val excludedExpenseMinor: Halalas,
)

/** مجاميع الفترة. المدخل عمليات الفترة بس — الدالة ما بتفلترش بالتاريخ. */
fun computePeriodTotals(transactions: List<Transaction>, allocations: List<PersonAllocation>): PeriodTotals {
    var income = 0L
    var expense = 0L
    var excluded = 0L
    for (t in transactions) {
        if (countsAsIncome(t.economicKind)) {
            income = addMoney(income, t.amountMinor)
            continue
        }
        if (!countsAsPersonalExpense(t.economicKind)) continue
        val share = personalShareOf(t, allocations)
        if (t.excludedFromBudget) excluded = addMoney(excluded, share) else expense = addMoney(expense, share)
    }
    return PeriodTotals(income, expense, subtractMoney(income, expense), excluded)
}

/** مجموع العمليات الموسومة — العمليات **المميزة** بمعرّفها، فوسمين على عملية واحدة = مبلغها مرة واحدة. */
fun sumByTag(transactions: List<Transaction>, taggedTransactionIds: List<String>, allocations: List<PersonAllocation> = emptyList()): Halalas {
    val wanted = taggedTransactionIds.toSet()
    val seen = mutableSetOf<String>()
    var total = 0L
    for (t in transactions) {
        if (t.id !in wanted || !seen.add(t.id)) continue
        total = addMoney(total, personalShareOf(t, allocations))
    }
    return total
}

data class PersonBalance(
    val personId: String,
    /** «لك عنده». */
    val receivableMinor: Halalas,
    /** «له عندك» — قروض. */
    val payableLoanMinor: Halalas,
    /** «له عندك» — أمانات. ما بتتقاصش مع القرض. */
    val payableCustodyMinor: Halalas,
)

/** المتبقي من التزام = الأصل − تسوياته. السالب ممنوع ومش بيتسكت عليه. */
fun remainingOfObligation(obligation: Obligation, settlements: List<Settlement>): Halalas {
    val paid = sumMoney(settlements.filter { it.obligationId == obligation.id }.map { it.amountMinor })
    val remaining = subtractMoney(obligation.originalMinor, paid)
    if (remaining < 0) throw LedgerError(uiText(TextKey.LEDGER_SETTLEMENTS_EXCEED, obligation.id))
    return remaining
}

/** أرصدة شخص — **مفيش تقاص تلقائي** بين «لك» و«عليك»، والقرض والأمانة منفصلين (spec/02). */
fun computePersonBalance(personId: String, obligations: List<Obligation>, settlements: List<Settlement>): PersonBalance {
    val mine = obligations.filter { it.personId == personId }
    fun totalOf(kind: ObligationKind) = sumMoney(mine.filter { it.kind == kind }.map { remainingOfObligation(it, settlements) })
    return PersonBalance(personId, totalOf(ObligationKind.RECEIVABLE), totalOf(ObligationKind.LOAN_PAYABLE), totalOf(ObligationKind.CUSTODY_PAYABLE))
}

data class SettlementCheck(
    val allowed: Boolean,
    /** سبب الرفض بلغة المستخدم (spec/06). */
    val reason: String?,
    /** اللي بيتسوّى فعلًا. */
    val settledMinor: Halalas,
    /** الزيادة اللي محتاجة إجراء صريح كأمانة مستقلة. */
    val surplusMinor: Halalas,
)

/** فحص تسوية قبل تطبيقها — الزيادة ما بتتبلعش ولا بتتسحب من التزام تاني (spec/06). */
fun checkSettlement(obligation: Obligation, settlements: List<Settlement>, amountMinor: Halalas): SettlementCheck {
    if (amountMinor <= 0) return SettlementCheck(false, uiText(TextKey.SETTLEMENT_AMOUNT_POSITIVE), 0, 0)
    val remaining = remainingOfObligation(obligation, settlements)
    if (remaining == 0L) return SettlementCheck(false, uiText(TextKey.SETTLEMENT_FULLY_PAID), 0, amountMinor)
    if (amountMinor <= remaining) return SettlementCheck(true, null, amountMinor, 0)
    val kindLabel = when (obligation.kind) {
        ObligationKind.CUSTODY_PAYABLE -> uiText(TextKey.OBLIGATION_CUSTODY)
        ObligationKind.LOAN_PAYABLE -> uiText(TextKey.OBLIGATION_LOAN)
        ObligationKind.RECEIVABLE -> uiText(TextKey.OBLIGATION_RECEIVABLE)
    }
    val surplus = subtractMoney(amountMinor, remaining)
    return SettlementCheck(
        allowed = false,
        // بالريال مش بالهللة: الوحدة الداخلية ما تظهرش للمستخدم أبدًا
        reason = uiText(TextKey.SETTLEMENT_OVER_REMAINING, kindLabel, formatMoney(remaining), formatMoney(amountMinor), formatMoney(surplus)),
        settledMinor = remaining,
        surplusMinor = surplus,
    )
}

data class GrossSplit(val principalMinor: Halalas, val feeMinor: Halalas)

/** «رسوم 20 ضمن خصم 1020 ⇒ أصل 1000 + رسوم 20؛ مش 1040» (spec/06). */
fun splitGrossIntoPrincipalAndFee(grossMinor: Halalas, feeMinor: Halalas): GrossSplit {
    if (feeMinor < 0) throw LedgerError(uiText(TextKey.LEDGER_FEE_NEGATIVE))
    if (feeMinor > grossMinor) throw LedgerError(uiText(TextKey.LEDGER_FEE_OVER_GROSS, feeMinor.toString(), grossMinor.toString()))
    return GrossSplit(subtractMoney(grossMinor, feeMinor), feeMinor)
}
